### DISCUSSION TITLE
v0.7.x Breaks AWS HTTP Lambda Example

### DISCUSSION THREAD

**Discussion by davidachris** [tokio-rs/axum#2372] [Q&A]
- [x] I have looked for existing issues (including closed) about this

### Version
v0.7.x

### Crates

```
[dependencies]
axum = "0.7"
lambda_http = "0.8"
lambda_runtime = "0.8"
serde_json = "1"
serde = { version = "1", features = ["derive"] }
tokio = { version = "1", features = ["macros", "full"] }
tracing = { version = "0.1", features = ["log"] }
tracing-subscriber = { version = "0.3", default-features = false, features = [
    "fmt",
] }
```

### Description
In summary, the AWS Lambda example we could previously use Axum up to v6.x. After upgrading to v7.x, I get errors because of possibly an incompatibility. I was able to track it down to a removal of generics in the Router Struct. Please see below in breaking changes. Appologies if this is an intended change; however, I could not find information in the documentation in a different way to setup a Router so that it may be compatible with the AWS example, if its simply no longer possible to use in Lambda thats fine as well, I wanted to at least bring it to your attention. Thanks!

If you upgrade this example to v.7.x then this is no longer possible
https://github.com/awslabs/aws-lambda-rust-runtime/tree/main/examples/http-axum

v0.7.x:

    axum::routing
    pub struct Router<S = ()> {
        path_router: PathRouter<S, false>,
        fallback_router: PathRouter<S, true>,
        default_fallback: bool,
        catch_all_fallback: Fallback<S, Infallible>,
    }

v0.6.x:

    axum::routing
    pub struct Router<S = (), **B = Body**> {
        path_router: PathRouter<S, B, false>,
        fallback_router: PathRouter<S, B, true>,
        default_fallback: bool,
        catch_all_fallback: Fallback<S, B, Infallible>,
    }

v0.7.x <error msg>:

    error[E0277]: the trait bound `Router<_>: Service<lambda_http::http::Request<lambda_http::Body>>` is not satisfied
    --> src/main.rs:51:9
        |
    51  |     run(app).await
        |     --- ^^^ the trait `Service<lambda_http::http::Request<lambda_http::Body>>` is not implemented for `Router<_>`
        |     |
        |     required by a bound introduced by this call
        |
        = help: the following other types implement trait `Service<Request>`:
                <Router as Service<axum::http::Request<B>>>
                <Router as Service<IncomingStream<'_>>>
    note: required by a bound in `lambda_http::run`
    --> /Users/<user>/.cargo/registry/src/index.crates.io-6f17d22bba15001f/lambda_http-0.8.3/src/lib.rs:191:8
        |
    189 | pub async fn run<'a, R, S, E>(handler: S) -> Result<(), Error>
        |              --- required by a bound in this function
    190 | where
    191 |     S: Service<Request, Response = R, Error = E>,
        |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `run`

    For more information about this error, try `rustc --explain E0277`.

**Comment by davidpdrsn** [comment]
0.6 to 0.7 is a major version bump so it should not come as a surprise that things no longer compile.

axum 0.7 now requires the request body type to be `axum::body::Body`. If you have an existing type that implements `http_body::Body` (which you must have had for using axum 0.6) then you convert that in axum's body with [`Body::new`](https://docs.rs/axum/latest/axum/body/struct.Body.html#method.new).

**Reply by Severon96** [reply]
@davidpdrsn What would someone do, if he hasn't an implementing type of `http_body:Body`?

For example my current implementation looks like this:
```
#[tokio::main]
async fn main() -> Result<(), Error> {
    init_tracing();

    let connection = setup_db_connection();
    let api = setup_api();
    let app = Router::new().nest("/api", api)
        .with_state(connection);

    run(app).await
}

async fn welcome(State(pool): State<AsyncPool>) -> Result<Json<&'static str>, ServerError> {
    Ok(Json("Hello"))
}

fn setup_api() -> Router<Pool<AsyncPgConnection>> {
    return Router::new()
        .route("/", get(welcome));
}

fn setup_db_connection() -> Pool<AsyncPgConnection> {
    let db_url = std::env::var("DATABASE_URL").expect("missing DATABASE_URL environment variable");
    let config = AsyncDieselConnectionManager::<AsyncPgConnection>::new(db_url);

    return Pool::builder().build(config);
}

fn init_tracing() {
    tracing_subscriber::fmt()
        .with_max_level(tracing::Level::INFO)
        .with_target(false)
        .without_time()
        .init();
}
```

I receive the exact same error from above and I don't know what to adjust at my code to work with axum 0.7.

To be honest i'm starting getting into Rust, so i'm not that deep into the topic at the moment. Possible that i'm overseeing something obvious.

**Reply by Severon96** [reply]
Found a solution for me:
There where this issue in the AWS Lambda Rust Runtime which is already merged in main, but not released:
https://github.com/awslabs/aws-lambda-rust-runtime/issues/738

I could fix it in my project by depending on the master branch of the AWS Lambda Rust Runtime for lambda_http like this:
`lambda_http = { git = "https://github.com/awslabs/aws-lambda-rust-runtime", branch = "main" }`

As soon as there is a new release this can be replaced by the regular version number.
